### ISSUE TITLE
IPC barrier: returned gradient `dbdd` is not the derivative of the returned barrier `b` (off by `(d − d̂) + 1`), creating a spurious attraction band

### ISSUE THREAD

**Issue by transfix** [SetasAditya/material-aware-grl-snam#2]
### Summary

In the IPC-style obstacle barrier, the returned gradient `dbdd` is **not** the derivative of the returned barrier value `b`. The two disagree by exactly `(d − d̂) + 1`, which puts a spurious **attraction band** just outside each obstacle (the barrier pulls *toward* the obstacle over an outer band) and means the force field is not the gradient of any potential.

### Where

The buggy analytic derivative appears verbatim at three definition sites:

- `full_code/surrogate_robust.py:48` (`ipc_piecewise`)
- `full_code/train_coef_energy.py:112`
- `full_code/scripts/spline_stagewise6.py:548`

Each computes:

```python
b_in    = -(d - dh) ** 2 * torch.log(safe / dh)
dbdd_in =  (dh - d) * (2.0 * torch.log(safe / dh) - dh / safe) + 1.0
```

### The math

For the stated barrier `b(d) = −(d − d̂)² · log(d/d̂)`, the true derivative is

```
b'(d) = 2(d̂ − d)·log(d/d̂) − (d − d̂)²/d
```

The implemented `dbdd_in` expands to `2(d̂ − d)·log(d/d̂) − d̂(d̂ − d)/d + 1`, so

```
dbdd_in − b'(d) = (d − d̂) + 1      (exactly, for all d in (0, d̂))
```

Two consequences:

1. **Spurious attraction band.** Near obstacle-detection range the true derivative → 0, but the implemented one → 1, and it changes sign over an outer band (roughly `(0.39·d̂, d̂)` for small `d̂`, wider as `d̂` grows). There, `F_bar = −(α · dbdd) · n̂` points **toward** the obstacle instead of away from it.
2. **Non-conservative force.** Because `dbdd ≠ b′` (and the returned `b` value is computed but discarded — only `dbdd` feeds the force), the resulting force field is not `−∇` of `b`, or of any potential. This weakens any energy/least-action interpretation of the rollout.

### Note on the highway-env variant (not the same bug)

`full_code/exp-highway-env/surrogate_integrator.py:78-80` uses a *different but self-consistent* convention — there `db_dd` genuinely equals `d/dd` of its own `b_val` (scaled by `-vp`), and its docstring writes `b'(d)` correctly. So that file isn't wrong, but it uses a different scaling/sign convention than the three files above; worth reconciling everything to one convention while fixing.

### Suggested fix

Replace the analytic derivative at the three sites with the true `b′`, e.g.

```python
dbdd_in = (dh - d) * 2.0 * torch.log(safe / dh) - (d - dh) ** 2 / safe
```

(or compute it via autograd of `b_in`), and add a regression test asserting `dbdd == d/dd b` — a finite-difference / `torch.autograd.grad` check of `dbdd` against `b`, plus an assertion that there is no sign flip / attraction band anywhere in `(0, d̂)`. There is currently no test guarding this property.

For reference, the corresponding barrier in the upstream `CVC-Lab/GRL-SNAM` implementation was corrected to the true derivative and locked in with a gradcheck test, so the fix and its test pattern already exist there to mirror.

### Impact

Every trajectory that ever had an obstacle within `d̂` is affected, so any checkpoints trained against the current barrier would need re-training after the fix (the corrected barrier is steeper near contact — a small learning-rate retune may help, as it did upstream).
